File: swift/Sources/MLXAudioModels/Whisper/Attention.swift
// Attention.swift
// Multi-head attention with KV caching for Whisper.

import Foundation
import MLX
import MLXNN

/// Multi-head attention with optional KV caching.
///
/// Supports both self-attention and cross-attention modes.
/// KV caching is used for efficient autoregressive decoding.
public class MultiHeadAttention: Module {

    /// Hidden dimension.
    let nState: Int

    /// Number of attention heads.
    let nHead: Int

    /// Dimension per head.
    let headDim: Int

    /// Attention scaling factor.
    let scale: Float

    /// Query projection.
    @ModuleInfo var query: Linear

    /// Key projection (no bias).
    @ModuleInfo var key: Linear

    /// Value projection.
    @ModuleInfo var value: Linear

    /// Output projection.
    @ModuleInfo var out: Linear

    /// Initialize multi-head attention.
    ///
    /// - Parameters:
    ///   - nState: Hidden dimension (must be divisible by nHead)
    ///   - nHead: Number of attention heads
    public init(nState: Int, nHead: Int) {
        self.nState = nState
        self.nHead = nHead
        self.headDim = nState / nHead
        self.scale = Float(pow(Double(headDim), -0.5))

        // Separate projections for Q, K, V (Whisper style)
        self._query.wrappedValue = Linear(nState, nState)
        self._key.wrappedValue = Linear(nState, nState, bias: false)
        self._value.wrappedValue = Linear(nState, nState)
        self._out.wrappedValue = Linear(nState, nState)
    }

    /// Forward pass with optional KV caching.
    ///
    /// - Parameters:
    ///   - x: Query input [B, T, D]
    ///   - xa: Key/value source for cross-attention [B, S, D].
    ///         If nil, performs self-attention.
    ///   - mask: Attention mask [T, S] or [B, T, S]
    ///   - kvCache: Cached (key, value) from previous steps for
    ///              incremental decoding. Only used for self-attention.
    /// - Returns: Tuple of output tensor [B, T, D] and updated KV cache
    public func callAsFunction(
        _ x: MLXArray,
        xa: MLXArray? = nil,
        mask: MLXArray? = nil,
        kvCache: (MLXArray, MLXArray)? = nil
    ) -> (MLXArray, (MLXArray, MLXArray)?) {
        let B = x.dim(0)
        let T = x.dim(1)

        // Compute query
        var q = query(x)

        var k: MLXArray
        var v: MLXArray
        var newKvCache: (MLXArray, MLXArray)?

        if let xa = xa {
            // Cross-attention (no caching needed, encoder output is fixed)
            k = key(xa)
            v = value(xa)
            newKvCache = nil
        } else {
            // Self-attention
            k = key(x)
            v = value(x)

            // Handle KV cache for incremental decoding
            if let (kCache, vCache) = kvCache {
                k = concatenated([kCache, k], axis: 1)
                v = concatenated([vCache, v], axis: 1)
            }

            newKvCache = (k, v)
        }

        let S = k.dim(1)

        // Reshape for multi-head attention
        // [B, T, D] -> [B, T, nHead, headDim] -> [B, nHead, T, headDim]
        q = q.reshaped([B, T, nHead, headDim]).transposed(0, 2, 1, 3)
        k = k.reshaped([B, S, nHead, headDim]).transposed(0, 2, 1, 3)
        v = v.reshaped([B, S, nHead, headDim]).transposed(0, 2, 1, 3)

        // Compute attention scores
        // [B, nHead, T, headDim] @ [B, nHead, headDim, S] -> [B, nHead, T, S]
        var attn = (q.matmul(k.transposed(0, 1, 3, 2))) * scale

        // Apply mask if provided
        if let mask = mask {
            attn = attn + mask
        }

        // Softmax and apply to values
        attn = softmax(attn, axis: -1)

        // [B, nHead, T, S] @ [B, nHead, S, headDim] -> [B, nHead, T, headDim]
        var output = attn.matmul(v)

        // Reshape back
        // [B, nHead, T, headDim] -> [B, T, nHead, headDim] -> [B, T, D]
        output = output.transposed(0, 2, 1, 3).reshaped([B, T, nState])

        // Output projection
        output = out(output)

        return (output, newKvCache)
    }
}

/// Residual attention block for Whisper transformer.
///
/// Structure:
/// - LayerNorm -> Multi-head Self-Attention -> Residual
/// - (Optional) LayerNorm -> Multi-head Cross-Attention -> Residual
/// - LayerNorm -> MLP -> Residual
public class ResidualAttentionBlock: Module {

    /// Hidden dimension.
    let nState: Int

    /// Number of attention heads.
    let nHead: Int

    /// Whether this block has cross-attention.
    let hasCrossAttention: Bool

    /// Self-attention layer.
    @ModuleInfo var attn: MultiHeadAttention

    /// Layer norm for self-attention.
    @ModuleInfo var attnLn: LayerNorm

    /// Cross-attention layer (decoder only).
    @ModuleInfo var crossAttn: MultiHeadAttention?

    /// Layer norm for cross-attention.
    @ModuleInfo var crossAttnLn: LayerNorm?

    /// First linear layer of MLP.
    @ModuleInfo var mlp0: Linear

    /// Second linear layer of MLP.
    @ModuleInfo var mlp1: Linear

    /// Layer norm for MLP.
    @ModuleInfo var mlpLn: LayerNorm

    /// Initialize residual attention block.
    ///
    /// - Parameters:
    ///   - nState: Hidden dimension
    ///   - nHead: Number of attention heads
    ///   - crossAttention: Whether to include cross-attention (decoder)
    public init(nState: Int, nHead: Int, crossAttention: Bool = false) {
        self.nState = nState
        self.nHead = nHead
        self.hasCrossAttention = crossAttention

        // Self-attention
        self._attn.wrappedValue = MultiHeadAttention(nState: nState, nHead: nHead)
        self._attnLn.wrappedValue = LayerNorm(dimensions: nState)

        // Cross-attention (decoder only)
        if crossAttention {
            self._crossAttn.wrappedValue = MultiHeadAttention(nState: nState, nHead: nHead)
            self._crossAttnLn.wrappedValue = LayerNorm(dimensions: nState)
        }

        // MLP with 4x expansion
        self._mlp0.wrappedValue = Linear(nState, nState * 4)
        self._mlp1.wrappedValue = Linear(nState * 4, nState)
        self._mlpLn.wrappedValue = LayerNorm(dimensions: nState)
    }

    /// Forward pass.
    ///
    /// - Parameters:
    ///   - x: Input tensor [B, T, D]
    ///   - xa: Encoder output for cross-attention [B, S, D]
    ///   - mask: Causal attention mask
    ///   - kvCache: Cached (key, value) for incremental decoding
    /// - Returns: Tuple of output tensor [B, T, D] and updated KV cache
    public func callAsFunction(
        _ x: MLXArray,
        xa: MLXArray? = nil,
        mask: MLXArray? = nil,
        kvCache: (MLXArray, MLXArray)? = nil
    ) -> (MLXArray, (MLXArray, MLXArray)?) {
        var x = x

        // Self-attention
        let (attnOut, newKvCache) = attn(attnLn(x), mask: mask, kvCache: kvCache)
        x = x + attnOut

        // Cross-attention (if applicable)
        if hasCrossAttention, let crossAttn = crossAttn, let crossAttnLn = crossAttnLn, let xa = xa {
            let (crossOut, _) = crossAttn(crossAttnLn(x), xa: xa)
            x = x + crossOut
        }

        // MLP
        let mlpOut = mlp1(gelu(mlp0(mlpLn(x))))
        x = x + mlpOut

        return (x, newKvCache)
    }
}

File: python/mlx_audio/models/whisper/layers/attention.py
"""Multi-head attention with KV caching for Whisper."""

from __future__ import annotations

import math
from typing import TYPE_CHECKING

import mlx.core as mx
import mlx.nn as nn

if TYPE_CHECKING:
    pass


class MultiHeadAttention(nn.Module):
    """Multi-head attention with optional KV caching.

    Supports both self-attention and cross-attention modes.
    KV caching is used for efficient autoregressive decoding.

    Attributes:
        n_state: Hidden dimension
        n_head: Number of attention heads
        scale: Attention scaling factor
    """

    def __init__(self, n_state: int, n_head: int):
        """Initialize multi-head attention.

        Args:
            n_state: Hidden dimension (must be divisible by n_head)
            n_head: Number of attention heads
        """
        super().__init__()
        self.n_state = n_state
        self.n_head = n_head
        self.head_dim = n_state // n_head
        self.scale = self.head_dim ** -0.5

        # Separate projections for Q, K, V (Whisper style)
        self.query = nn.Linear(n_state, n_state)
        self.key = nn.Linear(n_state, n_state, bias=False)
        self.value = nn.Linear(n_state, n_state)
        self.out = nn.Linear(n_state, n_state)

    def __call__(
        self,
        x: mx.array,
        xa: mx.array | None = None,
        mask: mx.array | None = None,
        kv_cache: tuple[mx.array, mx.array] | None = None,
    ) -> tuple[mx.array, tuple[mx.array, mx.array] | None]:
        """Forward pass with optional KV caching.

        Args:
            x: Query input [B, T, D]
            xa: Key/value source for cross-attention [B, S, D].
                If None, performs self-attention.
            mask: Attention mask [T, S] or [B, T, S]
            kv_cache: Cached (key, value) from previous steps for
                incremental decoding. Only used for self-attention.

        Returns:
            Tuple of:
                - Output tensor [B, T, D]
                - Updated KV cache (key, value) for self-attention,
                  or None for cross-attention
        """
        B, T, _ = x.shape

        # Compute query
        q = self.query(x)

        if xa is None:
            # Self-attention
            k = self.key(x)
            v = self.value(x)

            # Handle KV cache for incremental decoding
            if kv_cache is not None:
                k_cache, v_cache = kv_cache
                k = mx.concatenate([k_cache, k], axis=1)
                v = mx.concatenate([v_cache, v], axis=1)

            new_kv_cache = (k, v)
        else:
            # Cross-attention (no caching needed, encoder output is fixed)
            k = self.key(xa)
            v = self.value(xa)
            new_kv_cache = None

        S = k.shape[1]

        # Reshape for multi-head attention
        # [B, T, D] -> [B, T, n_head, head_dim] -> [B, n_head, T, head_dim]
        q = q.reshape(B, T, self.n_head, self.head_dim).transpose(0, 2, 1, 3)
        k = k.reshape(B, S, self.n_head, self.head_dim).transpose(0, 2, 1, 3)
        v = v.reshape(B, S, self.n_head, self.head_dim).transpose(0, 2, 1, 3)

        # Compute attention scores
        # [B, n_head, T, head_dim] @ [B, n_head, head_dim, S] -> [B, n_head, T, S]
        attn = (q @ k.transpose(0, 1, 3, 2)) * self.scale

        # Apply mask if provided
        if mask is not None:
            attn = attn + mask

        # Softmax and apply to values
        attn = mx.softmax(attn, axis=-1)

        # [B, n_head, T, S] @ [B, n_head, S, head_dim] -> [B, n_head, T, head_dim]
        out = attn @ v

        # Reshape back
        # [B, n_head, T, head_dim] -> [B, T, n_head, head_dim] -> [B, T, D]
        out = out.transpose(0, 2, 1, 3).reshape(B, T, self.n_state)

        # Output projection
        out = self.out(out)

        return out, new_kv_cache


class ResidualAttentionBlock(nn.Module):
    """Residual attention block for Whisper transformer.

    Structure:
        - LayerNorm -> Multi-head Self-Attention -> Residual
        - (Optional) LayerNorm -> Multi-head Cross-Attention -> Residual
        - LayerNorm -> MLP -> Residual

    Attributes:
        attn: Self-attention layer
        cross_attn: Cross-attention layer (decoder only)
        mlp: Feed-forward network
    """

    def __init__(
        self,
        n_state: int,
        n_head: int,
        cross_attention: bool = False,
    ):
        """Initialize residual attention block.

        Args:
            n_state: Hidden dimension
            n_head: Number of attention heads
            cross_attention: Whether to include cross-attention (decoder)
        """
        super().__init__()
        self.n_state = n_state
        self.n_head = n_head

        # Self-attention
        self.attn = MultiHeadAttention(n_state, n_head)
        self.attn_ln = nn.LayerNorm(n_state)

        # Cross-attention (decoder only)
        self.cross_attention = cross_attention
        if cross_attention:
            self.cross_attn = MultiHeadAttention(n_state, n_head)
            self.cross_attn_ln = nn.LayerNorm(n_state)

        # MLP
        self.mlp = nn.Sequential(
            nn.Linear(n_state, n_state * 4),
            nn.GELU(),
            nn.Linear(n_state * 4, n_state),
        )
        self.mlp_ln = nn.LayerNorm(n_state)

    def __call__(
        self,
        x: mx.array,
        xa: mx.array | None = None,
        mask: mx.array | None = None,
        kv_cache: tuple[mx.array, mx.array] | None = None,
    ) -> tuple[mx.array, tuple[mx.array, mx.array] | None]:
        """Forward pass.

        Args:
            x: Input tensor [B, T, D]
            xa: Encoder output for cross-attention [B, S, D]
            mask: Causal attention mask
            kv_cache: Cached (key, value) for incremental decoding

        Returns:
            Tuple of:
                - Output tensor [B, T, D]
                - Updated KV cache (self-attention only)
        """
        # Self-attention
        attn_out, new_kv_cache = self.attn(
            self.attn_ln(x),
            mask=mask,
            kv_cache=kv_cache,
        )
        x = x + attn_out

        # Cross-attention (if applicable)
        if self.cross_attention and xa is not None:
            cross_out, _ = self.cross_attn(
                self.cross_attn_ln(x),
                xa=xa,
            )
            x = x + cross_out

        # MLP
        x = x + self.mlp(self.mlp_ln(x))

        return x, new_kv_cache
